src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -538,6 +538,12 @@ private static void fetchSymlinkDependenciesRecursively(
 
   private static void fetchSymlinkDependenciesRecursively(
       ActionGraph actionGraph, Set<ActionInput> builder, Artifact artifact) {
+    if (!(actionGraph.getGeneratingAction(artifact) instanceof ActionExecutionMetadata)) {
+      // The top-level artifact could be a tree artifact, in which case the generating action may
+      // be an ActionTemplate, which does not implement ActionExecutionMetadata. We don't handle
+      // this case right now, so exit.
```

Comment:
Is the concern here that an ActionTemplate might create symlinks? I think that currently ActionTemplate is only implemented by CppCompileActionTemplate and SpawnActionTemplate. SpawnActionTemplate has no callers in Bazel, so I think there's no deficiency here as it stands. If we ever allow Starlark actions to create ActionTemplates, then I think there will be a gap here.

Reply:
Yes. Agreed that current templates don't have this problem, but we expect to see it in the future when / if Starlark is allowed to create them or if we add more templates.